Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -1736,6 +1746,343 @@ RefPtr<StyleValue> Parser::parse_image_value(ParsingContext const& context, Styl
+    };
+
+    if (component_values.size() == 1) {
+        auto maybe_value = parse_css_value(context, PropertyID::BackgroundRepeat, component_values.first());
+        if (!maybe_value)
+            return nullptr;
+        auto value = maybe_value.release_nonnull();
+        if (!is_background_repeat(*value))
+            return nullptr;
+
+        if (is_directional_repeat(value)) {
+            auto value_id = value->to_identifier();
+            return BackgroundRepeatStyleValue::create(
+                IdentifierStyleValue::create(value_id == ValueID::RepeatX ? ValueID::Repeat : ValueID::NoRepeat),
+                IdentifierStyleValue::create(value_id == ValueID::RepeatX ? ValueID::NoRepeat : ValueID::Repeat));
+        }
+        return BackgroundRepeatStyleValue::create(value, value);
+    }
+
+    if (component_values.size() == 2) {
+        auto maybe_x_value = parse_css_value(context, PropertyID::BackgroundRepeatX, component_values[0]);
+        auto maybe_y_value = parse_css_value(context, PropertyID::BackgroundRepeatY, component_values[1]);
+        if (!maybe_x_value || !maybe_y_value)
+            return nullptr;
+
+        auto x_value = maybe_x_value.release_nonnull();
+        auto y_value = maybe_y_value.release_nonnull();
+        if (!is_background_repeat(x_value) || !is_background_repeat(y_value))
+            return nullptr;
+        if (is_directional_repeat(x_value) || is_directional_repeat(y_value))
+            return nullptr;
+        return BackgroundRepeatStyleValue::create(x_value, y_value);
+    }
+
+    // FIXME: Handle multiple sets of comma-separated values.
+    dbgln("CSS Parser does not yet support multiple comma-separated values for background-repeat.");
+    return nullptr;
+}
+
+RefPtr<StyleValue> Parser::parse_border_value(ParsingContext const& context, PropertyID property_id, Vector<StyleComponentValueRule> const& component_values)
+{
+    auto is_line_style = [](StyleValue const& value) -> bool {
+        switch (value.to_identifier()) {
+        case ValueID::Dotted:
+        case ValueID::Dashed:
+        case ValueID::Solid:
+        case ValueID::Double:
+        case ValueID::Groove:
+        case ValueID::Ridge:
+        case ValueID::None:
+        case ValueID::Hidden:
+        case ValueID::Inset:
+        case ValueID::Outset:
+            return true;
+        default:
+            return false;
+        }
+    };
+
+    auto is_line_width = [](StyleValue const& value) -> bool {
+        if (value.is_length())
+            return true;
+
+        // FIXME: Implement thin/medium/thick
+        switch (value.to_identifier()) {
+        case ValueID::None:
+            return true;
+        default:
+            return false;
+        }
+    };
+
+    if (component_values.size() > 3)
+        return nullptr;
+
+    Optional<NonnullRefPtr<StyleValue>> maybe_border_width;
+    Optional<NonnullRefPtr<StyleValue>> maybe_border_color;
+    Optional<NonnullRefPtr<StyleValue>> maybe_border_style;
```

Comment:
As above making those `RefPtr<>`s with `nullptr` representing the `sorry-ive-got-no-value` state seems to be the more Serenity-like approach.